File: api/src/main/java/org/delivery/api/common/api/Result.java
package org.delivery.api.common.api;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.delivery.common.error.ErrorCode;
import org.delivery.common.error.ErrorCodeIfs;

/*@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Result {

    private Integer resultCode;

    private String resultMessage;

    private String resultDescription;

    public static Result ok() {
        return Result.builder()
                .resultCode(ErrorCode.OK.getErrorCode())
                .resultMessage(ErrorCode.OK.getDescription())
                .resultDescription("성공")
                .build();
    }

    public static Result error(ErrorCodeIfs errorCodeIfs) {
        return Result.builder()
                .resultCode(errorCodeIfs.getErrorCode())
                .resultMessage(errorCodeIfs.getDescription())
                .resultDescription("에러")
                .build();
    }

    public static Result error(ErrorCodeIfs errorCodeIfs, Throwable ex) {
        return Result.builder()
                .resultCode(errorCodeIfs.getErrorCode())
                .resultMessage(errorCodeIfs.getDescription())
                .resultDescription(ex.getLocalizedMessage())    // 비추천
                .build();
    }

    public static Result error(ErrorCodeIfs errorCodeIfs, String message) {
        return Result.builder()
                .resultCode(errorCodeIfs.getErrorCode())
                .resultMessage(errorCodeIfs.getDescription())
                .resultDescription(message)
                .build();
    }
}*/

File: common/src/main/kotlin/org/delivery/common/api/Result.kt
package org.delivery.common.api

import org.delivery.common.error.ErrorCode
import org.delivery.common.error.ErrorCodeIfs

data class Result (
    val resultCode: Int ?= null,
    val resultMessage: String ?= null,
    val resultDescription: String ?= null,
) {

    companion object {
        // 자바로 컴파일 시 블록 내의 코드는 static 으로 인식
        @JvmStatic
        fun ok(): Result {
            return Result(
                resultCode = ErrorCode.OK.getErrorCode(),
                resultMessage = ErrorCode.OK.getDescription(),
                resultDescription = "성공"
            )
        }

        @JvmStatic
        fun error(
            errorCodeIfs: ErrorCodeIfs
        ): Result {
            return Result(
                resultCode = errorCodeIfs.getErrorCode(),
                resultMessage = errorCodeIfs.getDescription(),
                resultDescription = "에러"
            )
        }

        @JvmStatic
        fun error(
            errorCodeIfs: ErrorCodeIfs,
            throwable: Throwable
        ): Result {
            return Result(
                resultCode = errorCodeIfs.getErrorCode(),
                resultMessage = errorCodeIfs.getDescription(),
                resultDescription = throwable.localizedMessage
            )
        }

        @JvmStatic
        fun error(
            errorCodeIfs: ErrorCodeIfs,
            errorDescription: String
        ): Result {
            return Result(
                resultCode = errorCodeIfs.getErrorCode(),
                resultMessage = errorDescription
            )
        }

    }
}